crates/ruff/src/rules/flake8_pyi/rules/type_alias_naming.rs

```diff
@@ -0,0 +1,93 @@
+use rustpython_parser::ast::{Expr, ExprKind};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+
+use crate::checkers::ast::Checker;
+
+#[violation]
+pub struct SnakeCaseTypeAlias {
+    pub name: String,
+}
+
+impl Violation for SnakeCaseTypeAlias {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { name } = self;
+        format!("Type alias `{name}` should be CamelCase")
+    }
+}
+
+#[violation]
+pub struct TSuffixedTypeAlias {
+    pub name: String,
+}
+
+impl Violation for TSuffixedTypeAlias {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { name } = self;
+        format!("Private type alias `{name}` should not be suffixed with `T` (the `T` suffix implies that an object is a TypeVar)")
+    }
+}
+
+fn is_t_suffixed_type_alias(name: &str) -> bool {
+    // A T suffixed, private type alias must begin with an underscore.
+    if !name.starts_with('_') {
+        return false;
+    }
+
+    // It must end in a lowercase letter, followed by `T`, and (optionally) a digit.
+    let mut revchars = name.chars().rev();
+    matches!(
+        (revchars.next(), revchars.next(), revchars.next()), // note this is end -> beginning
+        (Some('0'..='9'), Some('T'), Some('a'..='z')) | (Some('T'), Some('a'..='z'), _)
+    )
+}
+
+fn is_snake_case_type_alias(name: &str) -> bool {
+    // The first letter must be alphabetic, optionally preceded by an '_'.
+    let mut chars = name.chars();
+    let Some(mut first) = chars.next() else { return false; };
+
+    if first == '_' {
+        if let Some(c) = chars.next() {
+            first = c;
+        } else {
+            return false;
+        }
+    }
+
+    // There should be at least one other underscore in the name.
+    first.is_ascii_alphabetic() && chars.any(|c| c == '_')
+}
```

Comment:
I did some testing against `typeshed`, and I think this matches their current behavior:

```rust
/// Return `true` if the given name is a snake_case type alias. In this context, we match against
/// any name that begins with an optional underscore, followed by at least one lowercase letter.
fn is_snake_case_type_alias(name: &str) -> bool {
    let mut chars = name.chars();
    match (chars.next(), chars.next()) {
        (Some('_'), Some(c)) | (Some(c), ..) => c.is_ascii_alphabetic() && c.is_ascii_lowercase(),
        _ => false,
    }
}
```

Reply:
The current version seems to have some false positives, e.g., it flags this, which `flake8-pyi` allows:

```py
_TYPE_FIELDS: TypeAlias = _TYPE_FIELDS_SEQUENCE | Mapping[str, fields._FieldValueTuple]
```

Reply:
That makes sense- and is the original behavior the regex supported (`^_?[a-z]`).

I changed it to support `Snake_case`- is it more important to maintain parity with the original plug-in?

Reply:
Ah yeah, I prefer to maintain parity with upstream plugins unless we have a good reason to deviate (examples: there's a clear bug that we can fix, diagnostic messages are confusing or inconsistent with our own conventions). It makes migrations easier for users, and reduces confusion.